### PR TITLE
http: add core retry policy to route retry policy conversion utility

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

Commit Message: The conversion from `core::RetryPolicy` to `route::RetryPolicy` is essential for basic stream option setup. But in current impl, that is isolated on JWK fetcher and needed to use https://github.com/envoyproxy/envoy/pull/17469. This refactor includes to the solution of them.
Additional Description:
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: envoy/http/async_client.h
Comment (mathetake) [inline envoy/http/async_client.h:229]: retry_on - can be const ref?

File: envoy/http/async_client.h
Comment (mathetake) [inline envoy/http/async_client.h:333]: same here

File: envoy/http/async_client.h
Comment (Shikugawa) [reply envoy/http/async_client.h]: Yes no need to use rval ref here.

File: envoy/http/async_client.h
Comment (mathetake) [inline envoy/http/async_client.h:228]: better add comment about the difference from the above `setRetryPolicy(const envoy::config::route::v3::RetryPolicy& p)`?

**Review by mathetake — COMMENTED** [review]
overall looks good to me.

File: envoy/http/async_client.h
Comment (lizan) [inline envoy/http/async_client.h:21]: nit: constexpr implies static
```suggestion
constexpr uint32_t RetryInitialDelayMilliseconds = 1000;
```

**Comment by lizan** [comment]
/assign-from @envoyproxy/maintainers

for non-Tetrands review

File: envoy/http/async_client.h
Comment (Shikugawa) [reply envoy/http/async_client.h:242]: we can't raise any exception in `envoy/`. But after we moved this to external utility. we can achieve that.

Comment (RyanTheOptimist) [reply envoy/http/async_client.h:242]: I don't notice any change to move to an external utility in this PR. Are you intending to do so?

Comment (Shikugawa) [reply envoy/http/async_client.h:242]: Yes. I will submit changes within today.

Comment (RyanTheOptimist) [reply envoy/http/async_client.h:242]: Excellent! Thanks.

/wait

File: source/common/http/utility.cc
Comment (RyanTheOptimist) [inline source/common/http/utility.cc:1040]: How bout using std::chrono::microseconds(1000) so that the units are part of the type. I think this will also avoid a subsequent conversion down on line 1061/1063.

auto base_interval = std::chrono::microseconds(1000)
auto max_interval = base_interval * 10;

**Review by RyanTheOptimist — COMMENTED** [review]
Two small comments, but looks good otherwise.

File: source/common/http/utility.cc
Comment (Shikugawa) [reply source/common/http/utility.cc:1040]: I think that we don't have utility API to convert `chrono::milliseconds` to `protobuf::Duration`. But we have `MillisecondsToDuration` that converts `uint64_t` to `protobuf::Duration`. I think this is the easiest way.

Comment (RyanTheOptimist) [reply source/common/http/utility.cc:1040]: Oh, that's annoying! :) But fair enough.

**Review by RyanTheOptimist — APPROVED** [review]
Looks good to me. Over to the next reviewer.

**Comment by Shikugawa** [comment]
@jmarantz Could you take a look?

File: source/common/http/utility.cc
Comment (lizan) [inline source/common/http/utility.cc:1041]:
```suggestion
  constexpr uint64_t base_interval_ms = 1000;
  constexpr uint64_t max_interval_ms = 10 * base_interval_ms;
```

**Comment by Shikugawa** [comment]
@lizan Could you take a look?

**Review by lizan — DISMISSED** [review]

**Comment by lizan** [comment]
@jmarantz mind doing a non-tetrands check?

File: source/common/http/utility.cc
Comment (jmarantz) [inline source/common/http/utility.cc:1054]: Is there a test that hits this exception?

File: test/common/http/utility_test.cc
Comment (jmarantz) [inline test/common/http/utility_test.cc:999]: is this the test for that throw? I think there's a variant of EXPECT_THROW that looks for a particular message.

**Review by jmarantz — APPROVED** [review]
lgtm modulo one nit/question.

File: source/common/http/utility.cc
Comment (Shikugawa) [reply source/common/http/utility.cc:1054]: added test with message

File: source/extensions/filters/http/common/jwks_fetcher.cc
Comment (jmarantz) [inline source/extensions/filters/http/common/jwks_fetcher.cc:67]: explicit type here, per Envoy style.

File: test/common/http/utility_test.cc
Comment (jmarantz) [inline test/common/http/utility_test.cc:975]: const

File: test/common/http/utility_test.cc
Comment (jmarantz) [inline test/common/http/utility_test.cc:990]: const

File: test/common/http/utility_test.cc
Comment (jmarantz) [inline test/common/http/utility_test.cc:982]: use actual type

**Review by jmarantz — DISMISSED** [review]
just a few nits, but I'll pass off to senior maintainers.

/assign-from @envoyproxy/senior-maintainers

**Review by jmarantz — APPROVED** [review]
Sorry this has bounced back & forth a few times :). I think we are all good on this PR if we can coax it through CI. It's failing on 2 tests in the context of coverage, which I've seen fail on other PRs also.

**Pull request merged**
